File: scripts/ci/verify-container-release-policy.py
#!/usr/bin/env python3
"""Fail closed when HotM container publication drifts from release-only CalVer."""

from __future__ import annotations

import json
import re
import sys
from pathlib import Path


ROOT = Path(sys.argv[1]).resolve() if len(sys.argv) > 1 else Path(__file__).resolve().parents[2]
IMAGE_WORKFLOW = ROOT / ".gitea/workflows/publish-hotm-ui-image.yml"
CI_WORKFLOW = ROOT / ".gitea/workflows/ui-ci.yml"
PACKAGE_JSON = ROOT / "ui/package.json"
CALVER_RE = re.compile(r"^[0-9]{4}\.(?:[1-9]|1[0-2])\.[0-9]+$")


def workflow_header(text: str) -> str:
    return text.split("\njobs:", maxsplit=1)[0]


def job_block(text: str, name: str) -> str:
    match = re.search(
        rf"^  {re.escape(name)}:\s*$\n(?P<body>.*?)(?=^  [A-Za-z0-9_-]+:\s*$|\Z)",
        text,
        re.MULTILINE | re.DOTALL,
    )
    return match.group(0) if match else ""


def main() -> int:
    failures: list[str] = []
    try:
        image_workflow = IMAGE_WORKFLOW.read_text()
        ci_workflow = CI_WORKFLOW.read_text()
        declared_version = json.loads(PACKAGE_JSON.read_text())["version"]
    except (OSError, KeyError, json.JSONDecodeError) as error:
        print(f"container release policy check failed: {error}", file=sys.stderr)
        return 1

    image_header = workflow_header(image_workflow)
    image_job = job_block(image_workflow, "build-and-push")
    if re.search(r"^\s+branches:\s*", image_header, re.MULTILINE):
        failures.append("publish-hotm-ui-image.yml must not have a branch trigger")
    if "tags: ['v*']" not in image_header:
        failures.append("publish-hotm-ui-image.yml must retain the v* release trigger")
    if "if: startsWith(github.ref, 'refs/tags/v')" not in image_job:
        failures.append("HotM image publication must have a release-ref job guard")
    if "mode=rolling" in image_job or "sha-${" in image_job or "SHORT_SHA" in image_job:
        failures.append("HotM image publication must not emit rolling or commit tags")

    required_calver_fragments = (
        r"^v[0-9]{4}\.([1-9]|1[0-2])\.[0-9]+$",
        'VERSION="${GITHUB_REF_NAME#v}"',
        "DECLARED_VERSION=$(jq -er '.version' ui/package.json)",
        'if [ "${VERSION}" != "${DECLARED_VERSION}" ]; then',
    )
    for fragment in required_calver_fragments:
        if fragment not in image_job:
            failures.append(f"HotM image publication is missing CalVer control: {fragment}")

    required_tag_sets = (
        'UI_TAGS="${GHCR_UI}:latest,${GHCR_UI}:${VERSION},${GITEA_UI}:latest,${GITEA_UI}:${VERSION}"',
        'BUNDLE_TAGS="${GHCR_BUNDLE}:latest,${GHCR_BUNDLE}:${VERSION},${GITEA_BUNDLE}:latest,${GITEA_BUNDLE}:${VERSION}"',
    )
    for tag_set in required_tag_sets:
        if tag_set not in image_job:
            failures.append(f"HotM image publication is missing symmetric release tags: {tag_set}")

    if not CALVER_RE.fullmatch(str(declared_version)):
        failures.append(f"ui/package.json version is not YYYY.M.P CalVer: {declared_version}")

    if job_block(ci_workflow, "publish-dev"):
        failures.append("ui-ci.yml still defines retired branch container publication")
    proxy_release = job_block(ci_workflow, "publish-release")
    required_proxy_fragments = (
        "github.event_name == 'create'",
        "startsWith(github.ref, 'refs/tags/v')",
        r"^v[0-9]{4}\.([1-9]|1[0-2])\.[0-9]+$",
        "DECLARED_VERSION=$(jq -er '.version' ui/package.json)",
    )
    for fragment in required_proxy_fragments:
        if fragment not in proxy_release:
            failures.append(f"agent-proxy release is missing release control: {fragment}")
    if "sha-${" in proxy_release or "SHORT_SHA" in proxy_release:
        failures.append("agent-proxy release must not emit commit tags")

    if failures:
        print("HotM container release policy check failed.", file=sys.stderr)
        for failure in failures:
            print(f"- {failure}", file=sys.stderr)
        return 1

    print("HotM container release policy check passed")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/ci/verify-container-release-policy.mjs
#!/usr/bin/env node

import { readFileSync } from 'node:fs';
import { dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

const scriptDir = dirname(fileURLToPath(import.meta.url));
const root = process.argv[2] ? resolve(process.argv[2]) : resolve(scriptDir, '../..');
const imageWorkflowPath = resolve(root, '.gitea/workflows/publish-hotm-ui-image.yml');
const ciWorkflowPath = resolve(root, '.gitea/workflows/ui-ci.yml');
const desktopReleaseWorkflowPath = resolve(root, '.gitea/workflows/desktop-release.yml');
const publishDistWorkflowPath = resolve(root, '.gitea/workflows/publish-dist.yml');
const packageJsonPath = resolve(root, 'ui/package.json');
const calVer = /^[0-9]{4}\.(?:[1-9]|1[0-2])\.[0-9]+$/;

function workflowHeader(text) {
  return text.split('\njobs:', 1)[0];
}

function jobBlock(text, name) {
  const lines = text.split('\n');
  const start = lines.findIndex((line) => line === `  ${name}:`);
  if (start < 0) return '';
  let end = start + 1;
  while (end < lines.length && !/^  [A-Za-z0-9_-]+:\s*$/.test(lines[end])) end += 1;
  return lines.slice(start, end).join('\n');
}

const failures = [];
let imageWorkflow;
let ciWorkflow;
let desktopReleaseWorkflow;
let publishDistWorkflow;
let declaredVersion;
try {
  imageWorkflow = readFileSync(imageWorkflowPath, 'utf8');
  ciWorkflow = readFileSync(ciWorkflowPath, 'utf8');
  desktopReleaseWorkflow = readFileSync(desktopReleaseWorkflowPath, 'utf8');
  publishDistWorkflow = readFileSync(publishDistWorkflowPath, 'utf8');
  declaredVersion = JSON.parse(readFileSync(packageJsonPath, 'utf8')).version;
} catch (error) {
  console.error(`container release policy check failed: ${error.message}`);
  process.exit(1);
}

const imageHeader = workflowHeader(imageWorkflow);
const imageJob = jobBlock(imageWorkflow, 'build-and-push');
if (/^\s+branches:\s*/m.test(imageHeader)) failures.push('publish-hotm-ui-image.yml must not have a branch trigger');
if (!imageHeader.includes("tags: ['v*']")) failures.push('publish-hotm-ui-image.yml must retain the v* release trigger');
if (!imageJob.includes("if: startsWith(github.ref, 'refs/tags/v')")) failures.push('HotM image publication must have a release-ref job guard');
if (imageJob.includes('mode=rolling') || imageJob.includes('sha-${') || imageJob.includes('SHORT_SHA')) {
  failures.push('HotM image publication must not emit rolling or commit tags');
}

for (const fragment of [
  '^v[0-9]{4}\\.([1-9]|1[0-2])\\.[0-9]+$',
  'VERSION="${GITHUB_REF_NAME#v}"',
  "DECLARED_VERSION=$(jq -er '.version' ui/package.json)",
  'if [ "${VERSION}" != "${DECLARED_VERSION}" ]; then',
]) {
  if (!imageJob.includes(fragment)) failures.push(`HotM image publication is missing CalVer control: ${fragment}`);
}

for (const tagSet of [
  'UI_TAGS="${GHCR_UI}:latest,${GHCR_UI}:${VERSION},${GITEA_UI}:latest,${GITEA_UI}:${VERSION}"',
  'BUNDLE_TAGS="${GHCR_BUNDLE}:latest,${GHCR_BUNDLE}:${VERSION},${GITEA_BUNDLE}:latest,${GITEA_BUNDLE}:${VERSION}"',
]) {
  if (!imageJob.includes(tagSet)) failures.push(`HotM image publication is missing symmetric release tags: ${tagSet}`);
}

if (!calVer.test(String(declaredVersion))) failures.push(`ui/package.json version is not YYYY.M.P CalVer: ${declaredVersion}`);
if (jobBlock(ciWorkflow, 'publish-dev')) failures.push('ui-ci.yml still defines retired branch container publication');

const proxyRelease = jobBlock(ciWorkflow, 'publish-release');
for (const fragment of [
  "github.event_name == 'create'",
  "startsWith(github.ref, 'refs/tags/v')",
  '^v[0-9]{4}\\.([1-9]|1[0-2])\\.[0-9]+$',
  "DECLARED_VERSION=$(jq -er '.version' ui/package.json)",
]) {
  if (!proxyRelease.includes(fragment)) failures.push(`agent-proxy release is missing release control: ${fragment}`);
}
if (proxyRelease.includes('sha-${') || proxyRelease.includes('SHORT_SHA')) failures.push('agent-proxy release must not emit commit tags');

const createRefGuard = "github.event_name != 'create' || startsWith(github.ref, 'refs/tags/v')";
for (const jobName of ['quality-gate', 'agent-proxy-quality-gate', 'mocked-playwright-ci']) {
  if (!jobBlock(ciWorkflow, jobName).includes(createRefGuard)) {
    failures.push(`ui-ci.yml ${jobName} must skip non-tag create events`);
  }
}

const desktopReleaseGuard = "github.event_name == 'workflow_dispatch' || startsWith(github.ref, 'refs/tags/v')";
if (!jobBlock(desktopReleaseWorkflow, 'verify-tag').includes(desktopReleaseGuard)) {
  failures.push('desktop-release.yml must skip non-tag create events');
}

const publishDistGuard = "github.event_name == 'workflow_dispatch' || github.event_name == 'push' || startsWith(github.ref, 'refs/tags/v')";
if (!jobBlock(publishDistWorkflow, 'publish-dist').includes(publishDistGuard)) {
  failures.push('publish-dist.yml must skip non-tag create events');
}

if (failures.length > 0) {
  console.error('HotM container release policy check failed.');
  for (const failure of failures) console.error(`- ${failure}`);
  process.exit(1);
}

console.log('HotM container release policy check passed');
